packages/runtime/src/next/components/page.tsx

```diff
@@ -1,51 +1,82 @@
 'use client'
 
-import { Suspense, memo, useMemo } from 'react'
+import { Suspense, memo, useEffect } from 'react'
 
-import { RuntimeProvider } from '../../runtimes/react'
-import { Page as PageMeta } from '../../components/page'
-import { MakeswiftHostApiClient } from '../../api/react'
-import { MakeswiftPageSnapshot } from '../client'
+import { type CacheData } from '../../api/react'
+
+import { useStore } from '../../runtimes/react/hooks/use-store'
+import { useMakeswiftHostApiClient } from '../../runtimes/react/host-api-client'
+import { Page as PageComponent } from '../../components/page'
+import { type MakeswiftPageSnapshot, type MakeswiftPageDocument } from '../client'
+
+import * as ReactPage from '../../state/react-page'
+import { registerDocumentEffect, updateAPIClientCache } from '../../state/actions'
 
 export type PageProps = {
   snapshot: MakeswiftPageSnapshot
 }
 
+const isServer = typeof window === 'undefined'
+
+function useCacheData(cacheData: CacheData) {
+  const client = useMakeswiftHostApiClient()
+
+  if (isServer) {
+    client.makeswiftApiClient.dispatch(updateAPIClientCache(cacheData))
+  }
+
+  useEffect(() => {
+    client.makeswiftApiClient.dispatch(updateAPIClientCache(cacheData))
```

Comment:
Can you clarify what do you mean by "unconditionally"? It's dispatched only if `cacheData` (or the client) has changed, and we do want to update the cache in the Redux store every time it's updated from the outside.

Reply:
Effects will run at least once so when this component mounts it will dispatch the `updateAPIClientCache` action (twice in `StrictMode` during development). This might result in unnecessary JavaScript execution for live pages, which could impact performance, unless the reducer is smart enough to make this effectively a no-op and we're only paying the price of dispatching the action and reducers processing it but returning the current state (thus there are no re-renders).

Reply:
How would we dispatch it conditionally here? 

In this case, the reducer will make the dispatch of cache data identical to what's been prepopulated on the server a no-op, as it only fills in non-existent state: https://github.com/makeswift/makeswift/pull/836/files#diff-cbd5599cdef9ce1b5d413cd682e2d662f33071c2b64be862ecc43799bce436faR105

Reply:
I would perhaps have a thunk that checks if we're in editable mode or not before dispatching so that you can dispatch the thunk unconditionally and it would conditionally dispatch the action.

As for the cache data being identical, it still seems to create a new instance of the state's objects, eventually resulting in a re-render, if I'm reading things correctly.

Perhaps the ideal approach is not to have to sync the data in the first place and have it happen at the place the data is fetch. Unfortunately I don't think this is possible given the server-client boundary present in Next.js. If we fetched the data in the server and populated the API cache there and the only way to pass that to the client is via the `snapshot` prop, we have no choice but to perform an effect on a component to then pull the cache data out of the `snapshot` and insert it into the _client's_ API cache 🤷🏻‍♂️

Reply:
> As for the cache data being identical, it still seems to create a new instance of the state's objects, eventually resulting in a re-render, if I'm reading things correctly.

It does not:

<img width="491" alt="reducer" src="https://github.com/user-attachments/assets/6743ed7d-bbfd-42b5-87b2-fca2725a4b54">

Reply:
Ah, my bad